### ISSUE TITLE
NumPy should warn (eventually raise an error?) on comparisons between complex numbers

### ISSUE THREAD

**Issue by shoyer** [numpy/numpy#15981]
Python (version 3, at least) raises an error, but NumPy uses lexicographic order (see https://github.com/numpy/numpy/issues/11505 for discussion).

We should clean this up to match Python. This behavior is fine for sorting, but I don't think it makes sense for comparisons between numbers. I think most comparisons between complex numbers are probably latent bugs (as was the case where I discovered this).

My suggestion would be to start by issuing a DeprecationWarning, and then maybe make it a true error at some point in the distance future.

### Reproducing code example:

```python
>>> x = 1 + 1j
>>> y = 2 + 1j
>>> x < y
TypeError: '<' not supported between instances of 'complex' and 'complex'
>>> np.array(x) < np.array(y)
True
```

### Numpy/Python version information:

1.18.2 3.6.9 (default, Nov  7 2019, 10:44:02) 
[GCC 8.3.0]

**Comment by seberg** [comment]
Sounds like a reasonable thing to me to try, pending no bigger push-backs issues (which I do not suspect). A (maybe) similar sort being different from comparisons is that we sort NaN to the end of an array.

**Comment by shoyer** [comment]
I think it would also be fair to require setting a special option in `np.sort` to handle complex values, but I can imagine that might get more push-back.

**Comment by seberg** [comment]
True, although maybe a second step. Deprecating comparison (and min/max) this would also fix `https://github.com/numpy/numpy/issues/15630`. In many ways, I guess that is the same issue, since minimum and maximum should stay in sync with both comparisons and clip.

**Comment by charris** [comment]
We do define "sorting order", which may not be what people expect for things like complex. Perhaps we should be clearer in the documentation. It is reasonable the raise errors for comparison operators and such, search sorted may be a special case. I think the resolution should be driven by use cases, if any.

**Comment by eric-wieser** [comment]
If we wanted to deprecate sorting complex numbers, the motivation would typically be explained as _"the request is ambiguous, refuse the temptation to guess"_. However, if what we used to guess is really what the user wants, we need a concise way to spell it.

In python, that would be
```python
sorted(list_of_complex, key=lambda c: (c.real, c.imag))
```
Do we have an equivalent spelling in numpy today?

**Comment by eric-wieser** [comment]
Ah, I think that spelling is
```python
np.take_along_axis(c_arr, np.lexsort((c_arr.real, c_arr.imag), axis), axis)
```
which isn't too bad.

**Comment by seberg** [comment]
Indeed, lexsort is the way to spell it out, and probably fine in most cases, since it is not very common hopefully.
IT could be tedious in code that is generic to float or complex but requires this sort order. But I wonder if there is an actual likely usecase. And the fact that I doubt it, means that we might actually get away with it.

In any case, I think that comparisons and min/max have a higher priority than the sort order change, although if we consider it, doing it at the same time may be best.

**Comment by shoyer** [comment]
I could also imagine `np.sort(c_arr, key=(c_arr.real, c_arr.imag))` as sugar for `np.take_along_axis(c_arr, np.lexsort((c_arr.real, c_arr.imag), axis), axis)`

**Comment by shoyer** [comment]
Or maybe `np.sortby(c_arr, key=(c_arr.real, c_arr.imag))`

**Comment by eric-wieser** [comment]
Which also would presumably bring about the existence of `np.max_by`, `np.min_by`, `np.search_sorted_by`, etc.

**Comment by Qiyu8** [comment]
Is [np.isclose](https://github.com/numpy/numpy/issues/15959) a special complex comparison case? as @mruberry mentioned, a well-defined and widely accepted notion of saying "a is close to b" even when both a and b are complex, That is what Python's `cmath.isclose` did.

**Comment by mruberry** [comment]
That's true, in PyTorch we do plan to support complex isclose, following NumPy's and cmath's approach. We also plan to error on complex functions that logically depend on comparison operations (clamp, min, max, sort...), and on complex inputs to torch.sign. We do have requests for a complex sign function (`x/|x|`) and sorting complex numbers by absolute value, however, so we'll likely create new functions (e.g. `torch.sgn`, `torch.sortabs`) or add kwargs (e.g. `torch.sort(use_abs=True...)`) to let our users specify the behavior they'd like.

**Comment by seberg** [comment]
We also have the very strange `np.sort_complex` function already ;). Whatever its oddness, it might actually help a bit with migrating the hopefully very very few existing users of complex sorting. Assuming they know their array is complex.

**Comment by HansBrende** [comment]
To clarify one detail: Python 3 does not raise an error, but rather returns `NotImplemented` from the `complex` comparison methods. Which later turns into a `TypeError` after both operands return `NotImplemented`. Returning `NotImplemented` would be the correct approach here.
